### PR TITLE
🐛 Source Github: Fix error handling for 404 streams

### LINKED ISSUE #10555: GitHub source connector fails if "Projects" feature is disabled
## Environment
- **Airbyte version**:  0.35.27-alpha
- **OS Version / Instance**: Ubuntu 20.04
- **Deployment**: Kubernetes
- **Source Connector and version**: GitHub 0.2.19
- **Destination Connector and version**: N/A
- **Severity**: Medium
- **Step where error happened**: Sync job

## Current Behavior
When syncing from a GitHub repository where the "Projects" feature is disabled, the sync is aborted with error.

As a workaround, disabling the "projects" table does work. My org has this feature universally disabled, so disabling the table is a reasonable workaround for me.

## Expected Behavior
The source connector should gracefully ignore this and continue.

There appears to be error handling for this case, but it doesn't appear to be catching the problem.

https://github.com/airbytehq/airbyte/blob/f01b0c769f99b515394184ddb7eb0e1c99f2e748/airbyte-integrations/connectors/source-github/source_github/streams.py#L127-L130

## Logs

<details>
<summary>LOG</summary> 

```
2022-02-22 19:07:51 source > Syncing stream: projects 
2022-02-22 19:07:51 source > Undefined error while reading records: Projects are disabled for this repository
2022-02-22 19:07:51 source > Encountered an exception while reading stream SourceGithub
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 108, in read
    internal_config=internal_config,
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 141, in _read_stream
    for record in record_iterator:
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 185, in _read_incremental
    for record_counter, record_data in enumerate(records, start=1):
  File "/airbyte/integration_code/source_github/streams.py", line 251, in read_records
    sync_mode=sync_mode, cursor_field=cursor_field, stream_slice=stream_slice, stream_state=stream_state
  File "/airbyte/integration_code/source_github/streams.py", line 150, in read_records
    raise e
  File "/airbyte/integration_code/source_github/streams.py", line 97, in read_records
    yield from super().read_records(stream_slice=stream_slice, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 366, in read_records
    response = self._send_request(request, request_kwargs)
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 333, in _send_request
    return backoff_handler(user_backoff_handler)(request, request_kwargs)
  File "/usr/local/lib/python3.7/site-packages/backoff/_sync.py", line 94, in retry
    ret = target(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/backoff/_sync.py", line 94, in retry
    ret = target(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 297, in _send
    response.raise_for_status()
  File "/usr/local/lib/python3.7/site-packages/requests/models.py", line 960, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 410 Client Error: Gone for url: https://api.github.com/repos/my_org/my_repo/projects?per_page=100&state=all
2022-02-22 19:07:51 source > Finished syncing SourceGithub
2022-02-22 19:07:51 source > SourceGithub runtimes:

2022-02-22 19:07:51 source > 410 Client Error: Gone for url: https://api.github.com/repos/my_org/my_repo/projects?per_page=100&state=all
Traceback (most recent call last):
  File "/airbyte/integration_code/main.py", line 13, in <module>
    launch(source, sys.argv[1:])
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py", line 127, in launch
    for message in source_entrypoint.run(parsed_args):
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py", line 118, in run
    for message in generator:
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 112, in read
    raise e
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 108, in read
    internal_config=internal_config,
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 141, in _read_stream
    for record in record_iterator:
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 185, in _read_incremental
    for record_counter, record_data in enumerate(records, start=1):
  File "/airbyte/integration_code/source_github/streams.py", line 251, in read_records
    sync_mode=sync_mode, cursor_field=cursor_field, stream_slice=stream_slice, stream_state=stream_state
  File "/airbyte/integration_code/source_github/streams.py", line 150, in read_records
    raise e
  File "/airbyte/integration_code/source_github/streams.py", line 97, in read_records
    yield from super().read_records(stream_slice=stream_slice, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 366, in read_records
    response = self._send_request(request, request_kwargs)
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 333, in _send_request
    return backoff_handler(user_backoff_handler)(request, request_kwargs)
  File "/usr/local/lib/python3.7/site-packages/backoff/_sync.py", line 94, in retry
    ret = target(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/backoff/_sync.py", line 94, in retry
    ret = target(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 297, in _send
    response.raise_for_status()
  File "/usr/local/lib/python3.7/site-packages/requests/models.py", line 960, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 410 Client Error: Gone for url: https://api.github.com/repos/my_org/my_repo/projects?per_page=100&state=all
2022-02-22 19:08:22 destination > 2022-02-22 19:08:22 INFO i.a.i.b.FailureTrackingAirbyteMessageConsumer(close):63 - Airbyte message consumer: succeeded.
```

</details>

## Steps to Reproduce
1. Create a new GitHub source that includes a project with the "Projects" feature disabled.
2. Connect to a destination and include the "projects" source table.
3. Run it.

## Are you willing to submit a PR?
Yes, but I have a reasonable workaround at the moment, so this is low priority.

### LINKED ISSUE #10594: GitHub source connector fails when /contributors endpoint returns 404
## Environment
- **Airbyte version**:  0.35.27-alpha
- **OS Version / Instance**: Ubuntu 20.04
- **Deployment**: Kubernetes
- **Source Connector and version**: GitHub 0.2.19
- **Destination Connector and version**: N/A
- **Severity**: Medium
- **Step where error happened**: Sync job

## Current Behavior
The GitHub connector ran into a 404 error attempting to sync a mostly empty repo. The repo had only a single commit in it (e.g. `README.md`, template code, etc from a scaffolding tool). The commit was performed using an email address that is not associated with any GitHub account.

screenshot-20220223-183137

The connector hit a 404 error attempting to pull the `/contributors?per_page=100` request and aborts further actions.

The [API docs](https://docs.github.com/en/rest/reference/repos#list-repository-contributors) aren't particularly clear when the 404 would occur, I assumed only on a missing repository.

Annoyingly, I get the same response from this scaffolded repo as I get on a properly missing repo. When I run the same request against a purely empty repo, I successfully get a list of contributors. This may ultimately be a GitHub API bug. :confused:
```
$ curl -u "my_user:my_token"  -H "Accept: application/vnd.github.v3+json" https://api.github.com/repos/my_org/my_repo/collaborators
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest/reference/repos#list-repository-collaborators"
}
```

## Expected Behavior
The 404 error should be handled and the connector should continue as if no contributors are present.

## Logs

<details>
<summary>LOG</summary> 

```
2022-02-23 00:19:00 source > SourceGithub runtimes:

2022-02-23 00:19:00 source > Syncing stream: collaborators 
2022-02-23 00:19:00 source > Undefined error while reading records: Not Found
2022-02-23 00:19:00 source > Encountered an exception while reading stream SourceGithub
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 108, in read
    internal_config=internal_config,
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 141, in _read_stream
    for record in record_iterator:
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 213, in _read_full_refresh
    for record in records:
  File "/airbyte/integration_code/source_github/streams.py", line 150, in read_records
    raise e
  File "/airbyte/integration_code/source_github/streams.py", line 97, in read_records
    yield from super().read_records(stream_slice=stream_slice, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 366, in read_records
    response = self._send_request(request, request_kwargs)
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 333, in _send_request
    return backoff_handler(user_backoff_handler)(request, request_kwargs)
  File "/usr/local/lib/python3.7/site-packages/backoff/_sync.py", line 94, in retry
    ret = target(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/backoff/_sync.py", line 94, in retry
    ret = target(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 297, in _send
    response.raise_for_status()
  File "/usr/local/lib/python3.7/site-packages/requests/models.py", line 960, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://api.github.com/repos/my_org/my_repo/collaborators?per_page=100
2022-02-23 00:19:00 source > Finished syncing SourceGithub
2022-02-23 00:19:00 source > SourceGithub runtimes:

2022-02-23 00:19:00 source > 404 Client Error: Not Found for url: https://api.github.com/repos/my_org/my_repo/collaborators?per_page=100
Traceback (most recent call last):
  File "/airbyte/integration_code/main.py", line 13, in <module>
    launch(source, sys.argv[1:])
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py", line 127, in launch
    for message in source_entrypoint.run(parsed_args):
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py", line 118, in run
    for message in generator:
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 112, in read
    raise e
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 108, in read
    internal_config=internal_config,
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 141, in _read_stream
    for record in record_iterator:
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 213, in _read_full_refresh
    for record in records:
  File "/airbyte/integration_code/source_github/streams.py", line 150, in read_records
    raise e
  File "/airbyte/integration_code/source_github/streams.py", line 97, in read_records
    yield from super().read_records(stream_slice=stream_slice, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 366, in read_records
    response = self._send_request(request, request_kwargs)
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 333, in _send_request
    return backoff_handler(user_backoff_handler)(request, request_kwargs)
  File "/usr/local/lib/python3.7/site-packages/backoff/_sync.py", line 94, in retry
    ret = target(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/backoff/_sync.py", line 94, in retry
    ret = target(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 297, in _send
    response.raise_for_status()
  File "/usr/local/lib/python3.7/site-packages/requests/models.py", line 960, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://api.github.com/repos/my_org/my_repo/collaborators?per_page=100
2022-02-23 00:19:31 destination > 2022-02-23 00:19:31 INFO i.a.i.b.FailureTrackingAirbyteMessageConsumer(close):65 - Airbyte message consumer: succeeded.
```

</details>

## Steps to Reproduce
1. Create a new GitHub repo with a single commit using a dummy email address not associated with any GitHub account. (maybe?)
2. Set up the connector with a destination and include the `collaborators` table.
3. Run it.

## Are you willing to submit a PR?
Yes. Currently disabling the `collaborators` functionality to do an initial load of all our data. Can come back to this later.

### PR DESCRIPTION
**Pull request by grubberr**
Signed-off-by: Sergey Chvalyuk <grubberr@gmail.com>

## What
Fix regression for streams which generate 404
```
requests.exceptions.HTTPError: 410 Client Error: Gone for url: ...
```
## How
We need to get empty list for such streams

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [x] Create a non-forked branch based on this PR and test the below items on it
- [x] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [x] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [x] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [x] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [x] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: docs/integrations/sources/github.md [hunk 1]
```diff
@@ -95,7 +95,8 @@ Your token should have at least the `repo` scope. Depending on which streams you
 
 | Version | Date       | Pull Request | Subject                                                                                                      |
 |:--------|:-----------| :--- |:-------------------------------------------------------------------------------------------------------------|
-| 0.2.20  | 2022-02-16 | [10384](https://github.com/airbytehq/airbyte/pull/10384) | Add new stream `Deployments`, `ProjectColumns`, `PullRequestCommits`                                       |
+| 0.2.21  | 2022-03-05 | [10878](https://github.com/airbytehq/airbyte/pull/10878) | Fix error handling for 404 streams                                                                         |
```
Comment (bazarnov) [inline docs/integrations/sources/github.md:98]: nit: Fix error handling for unavailable streams with `404` status code.

**Review by bazarnov — APPROVED** [review]
LGTM. Please fix the unit_tests. Let's make the description in changelog more informative.

File: docs/integrations/sources/github.md (hunk 1 above)
Comment (grubberr) [reply docs/integrations/sources/github.md]: good point, fixed

File: airbyte-integrations/connectors/source-github/source_github/streams.py [hunk 2]
```diff
@@ -116,25 +116,23 @@ def read_records(self, stream_slice: Mapping[str, any] = None, **kwargs) -> Iter
...
-                # `Repositories` stream is not available for repositories not in an organization.
-                # Handle "404 Client Error: Not Found for url: https://api.github.com/orgs/<org_name>/repos?per_page=100" error.
-                error_msg = f"Syncing `Repositories` stream isn't available for organization `{stream_slice['organization']}`."
-            elif e.response.status_code == requests.codes.GONE and "/projects?" in error_msg:
+            elif e.response.status_code == requests.codes.GONE and isinstance(self, Projects):
                 # Some repos don't have projects enabled and we we get "410 Client Error: Gone for
                 # url: https://api.github.com/repos/xyz/projects?per_page=100" error.
                 error_msg = f"Syncing `Projects` stream isn't available for repository `{stream_slice['repository']}`."
-            elif e.response.status_code == requests.codes.NOT_FOUND and "/orgs/" in error_msg:
+            elif e.response.status_code == requests.codes.NOT_FOUND and isinstance(self, Organizations):
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-github/source_github/streams.py:123]: can't we make this big block much smaller if we have a list of entities which could return 404 and use the same behavior when the 404 is for one of these entities? 

e.g: 
```
OPTIONAL_ENTITIES = [Repositories, Users, Organizations, ...]
....
if 404 and optional_entity
  handle
else: 
  raise
```

Comment (grubberr) [reply airbyte-integrations/connectors/source-github/source_github/streams.py:123]: Improved!

**Review by sherifnada — APPROVED** [review]
One last question about 404s. If the answer is no, then no changes are needed. If yes, could you implement the suggestion? No furhter review required

**Review by ChristopheDuong — APPROVED** [review]

**Comment by grubberr** [comment]
@sherifnada 

if customer provided next config:
```
{
    "repository": "user/repository"
}
```

but API only supports `organization/repository` ( not `user/repository`)
we can get 404 for such cases

Because we have a some number of issues with this 404 problem
I have re-implemented to catch 404 for ALL streams.

**Review by sherifnada — APPROVED** [review]

**Comment by midavadim** [comment]
@ChristopheDuong I saw you approved that review, but there were a few comments from you. I have replied to these comments. Could you please let me know if I can merge this PR into main?
Thank you!

**Pull request merged**